### PR TITLE
402 - Adding Initial Validation

### PR DESCRIPTION
**Pull request by codyseibert**
- Adding an initial useCase for validating the petition form information and displaying an alert containing a message for each field that failed.
- skipping irsNoticeFile in validation since it doesn't exist in the UI for now
- moving error notification to the top

### CODE REVIEW

**Review by dknesek — COMMENTED** [review]
One question, Cody. Can you explain the name "validatePetitionForm" (as opposed to "validatePetition")?

File: shared/src/business/useCases/validatePetitionForm.interactor.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+function PetitionFormValidator(rawResponse) {
+  Object.assign(this, rawResponse);
+}
+
+joiValidationDecorator(
+  PetitionFormValidator,
+  joi.object().keys({
+    caseType: joi.string().required(),
+    irsNoticeDate: joi.string().required(),
```
Comment (sadlerw) [inline shared/src/business/useCases/validatePetitionForm.interactor.js]: irsNoticeDate isn't required

File: shared/src/business/useCases/validatePetitionForm.interactor.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    // irsNoticeFile: joi.object().required(),
+    petitionFile: joi.object().required(),
+    procedureType: joi.string().required(),
+    trialLocation: joi.string().required(),
+  }),
+);
+
+/**
+ * uploadCasePdfs
```
Comment (sadlerw) [inline shared/src/business/useCases/validatePetitionForm.interactor.js]: validatePetitionForm?

File: web-client/src/views/FilePetition.jsx [hunk 3]
```diff
@@ -106,21 +107,24 @@ export default connect(
               </select>
             </div>
             <div role="listitem" className="usa-form-group">
-              <label htmlFor="procedure-type">4. Preferred trial city is</label>
+              <label htmlFor="trial-location">4. Preferred trial city is</label>
```
Comment (sadlerw) [inline web-client/src/views/FilePetition.jsx]: I changed all of these on that other PR

File: shared/src/business/useCases/validatePetitionForm.interactor.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+function PetitionFormValidator(rawResponse) {
+  Object.assign(this, rawResponse);
+}
+
+joiValidationDecorator(
+  PetitionFormValidator,
+  joi.object().keys({
+    caseType: joi.string().required(),
+    irsNoticeDate: joi.string().required(),
+    // irsNoticeFile: joi.object().required(),
```
Comment (kkoskelin) [inline shared/src/business/useCases/validatePetitionForm.interactor.js]: What's the plan with this line?  Are we expecting to revive this one shortly, or should it be removed from the code?

File: shared/src/business/useCases/validatePetitionForm.intteractor.test.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+const { validatePetitionForm } = require('./validatePetitionForm.interactor');
+
+describe('validatePetitionForm', () => {
```
Comment (kkoskelin) [inline shared/src/business/useCases/validatePetitionForm.intteractor.test.js]: This file has a typo in the filename "intteractor".

File: web-client/src/presenter/actions/validatePetitionFormAction.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+      petitionFile,
+    },
+  });
+
+  if (!errors) {
+    return path.success();
+  } else {
+    return path.error({
+      alertError: {
+        title: 'Fix the follow errors to submit your form.',
```
Comment (kkoskelin) [inline web-client/src/presenter/actions/validatePetitionFormAction.js]: `Fix the following...` perhaps?

File: web-client/src/presenter/actions/validatePetitionFormAction.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        messages: Object.keys(errors).map(key => errors[key]),
```
Comment (kkoskelin) [inline web-client/src/presenter/actions/validatePetitionFormAction.js]: Is it more direct to use `Object.values(errors)` here?

File: web-client/src/views/ErrorNotification.jsx [hunk 8]
```diff
@@ -30,7 +30,16 @@ class ErrorNotification extends React.Component {
...
             <div className="usa-alert-body">
               <h3 className="usa-alert-heading">{alertError.title}</h3>
-              <p className="usa-alert-text">{alertError.message}</p>
+              {alertError.message && (
+                <p className="usa-alert-text">{alertError.message}</p>
+              )}
+              {alertError.messages && (
+                <ul>
+                  {alertError.messages.map((message, idx) => (
+                    <ol key={idx}>- {message}</ol>
```
Comment (kkoskelin) [inline web-client/src/views/ErrorNotification.jsx]: I think maybe you meant to use `<li>` elements here.

File: web-client/src/views/FilePetition.jsx [hunk 9]
```diff
@@ -106,21 +107,24 @@ export default connect(
...
+                    value: e.target.value,
                   });
                 }}
               >
                 <option value=""> -- Select -- </option>
                 {trialCities.map((trialCity, idx) => (
-                  <option key={idx} value="{trialCity.city}, {trialCity.state}">
+                  <option
+                    key={idx}
+                    value={trialCity.city + ', ' + trialCity.state}
```
Comment (kkoskelin) [inline web-client/src/views/FilePetition.jsx]: Considering we're concatenating "city, state" three times, smells like a computed.

**Review by kkoskelin — CHANGES_REQUESTED** [review]
Awesome start.  A few things to change.

File: web-client/src/presenter/actions/validatePetitionFormAction.js (hunk 7 above)
Comment (codyseibert) [reply web-client/src/presenter/actions/validatePetitionFormAction.js]: Object.values isn't supported in IE11 it seems.

**Review by kkoskelin — APPROVED** [review]

**Pull request merged**